### PR TITLE
feat(5388): Allow multiple external URLs for SSO access

### LINKED ISSUE #5388: Multiple external URLs for SSO access
# Summary

Currently, only one `url` can be configured in ArgoCD ConfigMap (`argocd-cm.yaml`; see https://argoproj.github.io/argo-cd/operator-manual/declarative-setup/).
It should be possible to specify more than one valid url.

# Motivation

Our clusters generally have two access paths:
1. via F5 BigIP and
2. via a management network (directly to cluster via nginx Ingress).

This also means that we use 2 differing URLs to access these endpoints.

In general, the F5 path is for end users, whereas the management path is for system administrator use as a last resort mechanism in case the F5 path experiences a downtime. For failure analysis it is important to still be able to access ArgoCD (preferrable via GUI).

Currently, if trying to access ArgoCD via a URL which is not specified in `url` of ArgoCD ConfigMap the response is 
```
Invalid return_url
```

# Proposal

`url` in `argocd-cm.yaml` should accept a list of valid URLs which can be used for SSO.

### PR DESCRIPTION
**Pull request by ClifHouck**
My organization has interest in getting this functionality merged, so I'm taking up the mantle. I've re-based the original PR on current (79b18e03dc05f2d1da64d30650b12e064e6ef067) master. This currently builds, but tests do not pass. I'll be back online next week to fix tests, verify functionality of this change, and respond to any existing review comments from the original PR.

Preserving the original PR note here for reference:

With https://github.com/argoproj/argo-cd/pull/4780, were introduced security measures to ensure the return_url is pointing to the current ArgoCD instance.

In several occasions, an ArgoCD isntance could be exposed through multiple network connections. Internal addresses and restricted public addresses.

Currently, a single base URL can be configured in the the argocd configmap, preventing from exposing ArgoCD on several access paths.

This change allows to define multiple hosts on which ArgoCD can be exposed, keeping backward compatibility by adding a field additionalUrls accepting a list of additional URLS on which argoCD can be exposed

Fixes https://github.com/argoproj/argo-cd/issues/5388

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [x] Optional. My organization is added to USERS.md.
* [x] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

**Comment by agaudreault** [comment]
@ClifHouck make sure to adress the review in https://github.com/argoproj/argo-cd/pull/14208#pullrequestreview-2116512325. You can switch the PR to draft until the test are passing and everything is implemented. Thanks for taking up the mantle! 💪

File: util/oidc/oidc.go [hunk 1]
```diff
@@ -148,17 +148,21 @@ func NewClientApp(settings *settings.ArgoCDSettings, dexServerAddr string, dexTl
 	return &a, nil
 }
 
-func (a *ClientApp) oauth2Config(scopes []string) (*oauth2.Config, error) {
+func (a *ClientApp) oauth2Config(request *http.Request, scopes []string) (*oauth2.Config, error) {
```
Comment (ClifHouck) [inline util/oidc/oidc.go:151]: It seems like the request object is needed here. Re: https://github.com/argoproj/argo-cd/pull/14208/files#r1638640243

File: util/settings/settings.go [hunk 2]
```diff
@@ -1468,6 +1474,16 @@ func updateSettingsFromConfigMap(settings *ArgoCDSettings, argoCDCM *apiv1.Confi
 	if err := validateExternalURL(argoCDCM.Data[settingUiBannerURLKey]); err != nil {
 		log.Warnf("Failed to validate UI banner URL in configmap: %v", err)
 	}
+	if argoCDCM.Data[settingURLsKey] != "" {
+		if err := yaml.Unmarshal([]byte(argoCDCM.Data[settingURLsKey]), &settings.URLs); err != nil {
+			log.Warnf("Failed to decode all external URLs in configmap: %v", err)
```
Comment (ClifHouck) [inline util/settings/settings.go:1479]: Corrected messages here, re: https://github.com/argoproj/argo-cd/pull/14208/files#r1638642960

**Review by ClifHouck — COMMENTED** [review]
Regarding this last comment: https://github.com/argoproj/argo-cd/pull/14208/files#r1638649274 it seems like Dex is not using the new URLs setting. So I removed that addition.

**Comment by ClifHouck** [comment]
@agaudreault I think I've addressed the comments you had on the previous PR. Except this: 

> 
I think the feature should be documented in the ArgoCD https://argo-cd.readthedocs.io/en/stable/operator-manual/user-management/ documentation.

Agreed. I can add some documentation around this feature.

> Is url deprecated? are urls valid on their own or only additional urls? Is this feature supported by all identity providers, or only by OIDC?

@tjamet What do you think about the questions posed above?

**Comment by ClifHouck** [comment]
What do I need to about the license compliance failing? Is there a contributor agreement I need to sign?

**Comment by tjamet** [comment]
> @agaudreault I think I've addressed the comments you had on the previous PR. Except this: 
> 
> 
> 
> > I think the feature should be documented in the ArgoCD https://argo-cd.readthedocs.io/en/stable/operator-manual/user-management/ documentation.
> 
> 
> 
> Agreed. I can add some documentation around this feature.
> 
> 
> 
> > Is url deprecated? are urls valid on their own or only additional urls? Is this feature supported by all identity providers, or only by OIDC?
> 
> 
> 
> @tjamet What do you think about the questions posed above?

Hi!

Yes, it should be documented, happy to help doing so.

About the deprecation, this was my idea at the beginning but I faced some issues as `url` seemed used for more things. I will take a second look

**Comment by ClifHouck** [comment]
@tjamet Any updates?

**Review by agaudreault — CHANGES_REQUESTED** [review]

**Comment by tjamet** [comment]
> @tjamet Any updates?

Taking a second look, indeed, URL is used to register in dex, and the update to replace by the new proposed list is not that trivial.

Eventually, a decision to deprecate the url config in favour of a list could be taken in the future

File: docs/operator-manual/argocd-cm.yaml [hunk 3]
```diff
@@ -10,6 +10,10 @@ data:
   # Argo CD's externally facing base URL (optional). Required when configuring SSO
   url: https://argo-cd-demo.argoproj.io
 
+  # Additional externally facing base URLs (optional)
+  urls: |
```
[the remark this answers is not in the document]
Comment (ClifHouck) [reply docs/operator-manual/argocd-cm.yaml]: I think I've updated dex to at least load the additional URLs with the proper callback endpoint appended. 

I've also handled config reload when `settings.AdditionalURLs` is modified.

File: util/oidc/oidc.go [hunk 4]
```diff
@@ -148,17 +148,21 @@ func NewClientApp(settings *settings.ArgoCDSettings, dexServerAddr string, dexTl
...
 	endpoint, err := a.provider.Endpoint()
 	if err != nil {
 		return nil, err
 	}
+	redirectURL, err := a.settings.RedirectURLForRequest(request)
+	if err != nil {
+		redirectURL = a.redirectURI
```
[the remark this answers is not in the document]
Comment (ClifHouck) [reply util/oidc/oidc.go:159]: Added warning.

File: docs/operator-manual/argocd-cm.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (ClifHouck) [reply docs/operator-manual/argocd-cm.yaml]: Renamed.

File: docs/operator-manual/argocd-cm.yaml [hunk 5]
```diff
@@ -10,6 +10,10 @@ data:
...
+    - https://argo-cd-demo2.argoproj.io
+
```
[the remark this answers is not in the document]
Comment (ClifHouck) [reply docs/operator-manual/argocd-cm.yaml]: Added some short documentation in the user-management doc

**Comment by ClifHouck** [comment]
I believe I've addressed everything brought up so far, and re-based on latest master e2c5a555ce559f7284428a73565c53bbc9480986 for good measure. Not sure what to do about the linter failing on files I haven't changed.

**Comment by ClifHouck** [comment]
Also need guidance on what to do about the license compliance check.

**Comment by agaudreault** [comment]
>Also need guidance on what to do about the license compliance check.

@ClifHouck  You can ignore that.

**Review by agaudreault — COMMENTED** [review]
LGTM. Added in my sprint to give it a try internally, for both dex and native oidc. I should be able to test tomorrow.

**Comment by ClifHouck** [comment]
> LGTM. Added in my sprint to give it a try internally, for both dex and native oidc. I should be able to test tomorrow.

Thank you!

File: docs/operator-manual/argocd-cm.yaml [hunk 6]
```diff
@@ -10,6 +10,10 @@ data:
   # Argo CD's externally facing base URL (optional). Required when configuring SSO
   url: https://argo-cd-demo.argoproj.io
 
+  # Additional externally facing base URLs (optional)
+  addtionalUrls: |
```
Comment (agaudreault) [inline docs/operator-manual/argocd-cm.yaml:14]:
```suggestion
  additionalUrls: |
```

**Comment by agaudreault** [comment]
@ClifHouck Have you tested this image on an actual deployment? I am getting constant `Invalid redirect URL: the protocol and host (including port) must match and the path must be within allowed URLs if provided`. Might be a configuration problem on my part, but everything looks as expected. I am testing with the native oidc auth.

File: util/settings/settings.go [hunk 7]
```diff
@@ -406,6 +410,8 @@ const (
 	settingServerPrivateKey = "tls.key"
 	// settingURLKey designates the key where Argo CD's external URL is set
 	settingURLKey = "url"
+	// settingURLsKey designates the key where Argo CD's external URLs is set
+	settingURLsKey = "urls"
```
Comment (agaudreault) [inline util/settings/settings.go:414]: Invalid key to load configuration

```suggestion
	// settingAdditionalUrlsKey designates the key where Argo CD's additional external URLs are set
	settingAdditionalUrlsKey = "additionalUrls"
```

**Comment by ClifHouck** [comment]
> @ClifHouck Have you tested this image on an actual deployment? I am getting constant `Invalid redirect URL: the protocol and host (including port) must match and the path must be within allowed URLs if provided`. Might be a configuration problem on my part, but everything looks as expected. I am testing with the native oidc auth.

No, but I thought the included additional tests of the original PR would've covered this case. I'll setup an environment to do more in-depth testing. I may not be able to complete this before leaving for vacation though. I'll be back online on August 12th.

**Review by agaudreault — CHANGES_REQUESTED** [review]
Using the correct value from config, it is working for dex+oidc and native oidc. 🎉

File: util/settings/settings.go [hunk 8]
```diff
@@ -1470,6 +1476,16 @@ func updateSettingsFromConfigMap(settings *ArgoCDSettings, argoCDCM *apiv1.Confi
 	if err := validateExternalURL(argoCDCM.Data[settingUiBannerURLKey]); err != nil {
 		log.Warnf("Failed to validate UI banner URL in configmap: %v", err)
 	}
+	if argoCDCM.Data[settingURLsKey] != "" {
+		if err := yaml.Unmarshal([]byte(argoCDCM.Data[settingURLsKey]), &settings.AdditionalURLs); err != nil {
+			log.Warnf("Failed to decode all external URLs in configmap: %v", err)
```
Comment (agaudreault) [inline util/settings/settings.go:1481]:
```suggestion
	if argoCDCM.Data[settingAdditionalUrlsKey] != "" {
		if err := yaml.Unmarshal([]byte(argoCDCM.Data[settingAdditionalUrlsKey]), &settings.AdditionalURLs); err != nil {
			log.Warnf("Failed to decode all additional URLs in configmap: %v", err)
```

**Review by agaudreault — APPROVED** [review]
Lgtm @ishitasequeira

**Review by ishitasequeira — APPROVED** [review]

**Review by ishitasequeira — APPROVED** [review]
Thanks @ClifHouck for the PR!! and Thanks @agaudreault for the review!!

**Review by alexmt — COMMENTED** [review]
LGTM

**Comment by akloss-cibo** [comment]
Any thoughts about when this will be in a tagged release?

**Comment by dave-gantenbein** [comment]
@ishitasequeira - we could really use a tagged release that contains this change, any idea when that might happen?

**Comment by blakepettersson** [comment]
This will be in `2.13`.

**Pull request merged**